dojo_plugin/api/v1/discord.py

```diff
@@ -56,5 +65,129 @@ def get(self, discord_id):
+    if end_stamp:
+        try:
+            end = datetime.fromisoformat(start_stamp)
+        except:
+            return {"success": False, "error": "invalid end format"}, 400
+
+    user = DiscordUsers.query.filter_by(discord_id=discord_id).first()
+
+    return get_user_activity_prop(discord_id, activity, start, end)
+
+def post_user_activity(discord_id, activity, request):
+    authorization = request.headers.get("Authorization")
+    res, code = auth_check(authorization)
+    if res:
+        return res, code
+
+    #data = request.get_json()
+    try:
+        data = json.loads(request.get_data())
+    except json.JSONDecodeError:
+        return {"success": False, "error": f"Invalid JSON data {request.data}"}, 400
+
+    if "from_user_id" not in data:
+        return {"success": False, "error": f"Invalid JSON data"}, 400
+
+    # These are discord user_ids
+    thanker = data.get("from_user_id", "")
+    kwargs = {'guild_id': data.get("guild_id"),
+              'channel_id': data.get("channel_id"),
+              'message_id': data.get("message_id"),
+              'timestamp': data.get("timestamp"),
+              'activity_type': activity
+              }
+    entry = DiscordUserActivity(discord_id, thanker, **kwargs)
+    db.session.add(entry)
+    db.session.commit()
+
+    return get_user_activity_prop(discord_id, activity)
+
+@discord_namespace.route("/memes/user/<discord_id>", methods=["GET", "POST"])
+class DiscordMemes(Resource):
+    def get(self, discord_id):
+      get_user_activity(discord_id, DiscordUserActivity.ActivityType.memes, request)
+
+    def post(self, discord_id):
+      post_user_activity(discord_id, DiscordUserActivity.ActivityType.memes, request)
+
+@discord_namespace.route("/thanks/user/<discord_id>", methods=["GET", "POST"])
+class DiscordThanks(Resource):
+    def get(self, discord_id):
+      get_user_activity(discord_id, DiscordUserActivity.ActivityType.thanks, request)
+
+    def post(self, discord_id):
+      post_user_activity(discord_id, DiscordUserActivity.ActivityType.thanks, request)
+
+
+@discord_namespace.route("/thanks/leaderboard", methods=["GET"])
+class GetDiscordLeaderBoard(Resource):
+    def get(self):
+        start_stamp = request.args.get("start")
+
+        def year_stamp():
+            year = datetime.datetime.now().year
+            return datetime.datetime(year, 1, 1)
+
+        try:
+            if start_stamp is None:
+                start = year_stamp()
+            else:
+                start = datetime.datetime.fromisoformat(start_stamp)
+        except:
+            return {"success": False, "error": "invalid start format"}, 400
+
+        thanks_scores = DiscordUserActivity.query.with_entities(DiscordUserActivity.to_user_id, db.func.count(DiscordUserActivity.to_user_id)
+          ).filter(and_(DiscordUserActivity.timestamp >= start),
+                   DiscordUserActivity.activity_type == DiscordUserActivity.ActivityType.memes
```

Comment:
@robwaz Should this be `DiscordUserActivity.ActivityType.thanks`?

Reply:
Slight addendum: Are we opposed to a memes leaderboard? Doesn't need to be part of this, but just wanted to ask.

Reply:
Good catch!  That's actually a fragment of a memes leaderboard I had.  I decided not to keep it because the memes leaderboard would need to not be limited to one meme per week. The backend already supports this.  However, from an impact side, I think encouraging quantity of memes over quality would lead to undesirable outcomes.